Upgrade Guide for Azure Landing Zone (#77)
Azure/Azure-Landing-Zones: docs/content/accelerator/startermodules/terraform-platform-landing-zone/Upgrade-guide.md

3. [Run terraform plan to see what changes will be made to your environment](#3-run-terraform-plan)

 ## 1. Review the release notes
The release notes will provide you with information on what has changed in the module, including any breaking changes, new features, and bug fixes. This will help you understand what to expect when upgrading to the latest version.

## 2. Update the module version
To ensure you are using the latest version of the module, you will need to update the version constraint in your Terraform configuration file. This will allow you to pin to a specific version of the module, or allow for automatic upgrades to the latest patch release.

### hub_and_spoke_vnet Module
To pin to a specific version of the module, you can use the following syntax:

```terraform
module "hub_and_spoke_vnet" {
  source  = "Azure/avm-ptn-alz-connectivity-hub-and-spoke-vnet/azurerm"
  version = "<version>" 

}
```

To allow automatic upgrades to the latest patch release, use the following version constraint syntax:

```terraform
module "hub_and_spoke_vnet" {
  source  = "Azure/avm-ptn-alz-connectivity-hub-and-spoke-vnet/azurerm"
  version = "~> 0.1.2"

> 

Reviewer: and again

## Staying up to date

> Deploying the latest version of the module is the recommended approach for staying up to date with the latest architectural changes to Azure landing zones. Looking at it from a governance perspective, this also ensures you have the latest recommended policies applied to your environment for improved compliance.

Reviewer 2: ```suggestion
Using the latest version of the ALZ Platform landing zone pattern modules is the recommended approach for staying up to date with the latest architectural changes. From a governance perspective, this also ensures you have the latest recommended policies applied to your environment.
```

> This upgrade guide covers four avm modules that are used to deploy the Azure platform landing zone. The modules are:

Reviewer 2: ```suggestion
This upgrade guide covers four AVM pattern modules that are used to deploy the Platform landing zone. The modules are:
```

> - [hub_and_spoke_vnet](hub_and_spoke_vnet)
> - [virtual_wan](virtual_wan)
> - [management_groups](management_groups)
> - [management_resources](management_resources)

Reviewer 2: ```suggestion
- [Management Groups and Policy](management_groups)
- [Management Resources](management_resources)
- [Connectivity with Hub and Spoke Virtual Network](hub_and_spoke_vnet)
- [Connectivity with Virtual WAN](virtual_wan)
```

> With each release of the module, it's possible that your environment will change. We will do our best to ensure any changes are clearly documented in the release notes, or upgrade guides when publishing a new major version. To avoid unexpected or unwanted changes we recommend that you configure your version constraints to pin to a specific module version.

Reviewer 2: ```suggestion
With each release of the AVM pattern modules, it's possible that there will be changes that could impact your deployed resources. We do our best to ensure any changes are documented in the release notes. To avoid unexpected or unwanted changes we recommend that you configure your version constraints to pin to a specific module version.
```

Upgrade process consists of the following steps:

> 1. [Review the release notes](#review-the-release-notes)

Reviewer 2: ```suggestion
1. [Review the release notes](#1-review-the-release-notes)
```

2. [Update the module version](#2-update-the-module-version)
3. [Run terraform plan to see what changes will be made to your environment](#3-run-terraform-plan)

>  ## 1. Review the release notes

Reviewer 2: ```suggestion
## 1. Review the release notes

```

The release notes will provide you with information on what has changed in the module, including any breaking changes, new features, and bug fixes. This will help you understand what to expect when upgrading to the latest version.

> ## 2. Update the module version

Reviewer 2: ```suggestion
## 2. Update the module version

```

To ensure you are using the latest version of the module, you will need to update the version constraint in your Terraform configuration file. This will allow you to pin to a specific version of the module, or allow for automatic upgrades to the latest patch release.

> ### hub_and_spoke_vnet Module

Reviewer 2: ```suggestion
### hub_and_spoke_vnet Module

```

To pin to a specific version of the module, you can use the following syntax:

```terraform
module "hub_and_spoke_vnet" {
  source  = "Azure/avm-ptn-alz-connectivity-hub-and-spoke-vnet/azurerm"
  version = "<version>" 
}
```

To allow automatic upgrades to the latest patch release, use the following version constraint syntax:

```terraform
module "hub_and_spoke_vnet" {
  source  = "Azure/avm-ptn-alz-connectivity-hub-and-spoke-vnet/azurerm"
  version = "~> 0.1.2"
}
```

> 
> 

Reviewer 2: ```suggestion

```

> ### Virtual_Wan Module

Reviewer 2: ```suggestion
### Virtual_Wan Module

```

To pin to a specific version of the module, you can use the following syntax:

> 

Reviewer 2: ```suggestion
```

```terraform
module "virtual_wan" {
  source  = "Azure/avm-ptn-alz-connectivity-virtual-wan/azurerm"
  version = "<version>" 
}
```

To allow automatic upgrades to the latest patch release, use the following version constraint syntax:

```terraform
module "virtual_wan" {
  source  = "Azure/avm-ptn-alz-connectivity-virtual-wan/azurerm"
  version = "~> 0.1.2"
}
```

###  Management Group Module
    
```terraform
module "management_groups" {

>     source     = "Azure/avm-ptn-alz/azurerm"

Reviewer 2: ```suggestion
  source  = "Azure/avm-ptn-alz/azurerm"
```

>     version = "<version>"

Reviewer 2: ```suggestion
  version = "<version>"
```

}
```

> To allow automatic upgrades to the latest patch release, use the following version constraint syntax:

Reviewer 2: ```suggestion

To allow automatic upgrades to the latest patch release, use the following version constraint syntax:
```

```terraform
module "management_groups" {
 source     = "Azure/avm-ptn-alz/azurerm"
 version     = "~> 0.16.0"
}

>  ```

Reviewer 2: ```suggestion
```
```

## 3.Run Terraform Plan and Apply

> ### Local file system

Reviewer 2: ```suggestion
### Local file system

```

The Terraform option outputs a `deploy-local.ps1` file that you can use to deploy the ALZ.
Follow the steps below to deploy the landing zone locally. If you want to hook it up to your custom version control system, follow their documentation on how to do that.
1. Navigate to the directory shown in the `module_output_directory_path` output from the bootstrap.
1. (Optional) Ensure you are still logged in to Azure using `az login --tenant 00000000-0000-0000-0000-000000000000`.
1. (Optional) Connect to your target subscription using `az account set --subscription 00000000-0000-0000-0000-000000000000`.
1. (Optional) Examine the `./scripts/deploy-local.ps1` to understand what it is doing.
1. Run `./scripts/deploy-local.ps1`.
1. A plan will run and then you'll be prompted to check it and run the deploy.
1. Type `yes` and hit enter to run the deploy.
1. The ALZ will now be deployed, this may take some time.

### Azure DevOps
1. In your Azure DevOps repository, create a new branch to test the upgrade. This ensures your changes are isolated and can be tested without affecting the main branch.

> 2. Modify the `version` field in Terraform module configuration to the desired version. Use version constraints to pin to a specific version or allow automatic upgrades to the latest patch release.

Reviewer 2: ```suggestion
1. Modify the `version` field in Terraform module configuration to the desired version. Use version constraints to pin to a specific version or allow automatic upgrades to the latest patch release.
```

> 3. Commit your changes to the new branch and update the pipeline to use the new branch.

Reviewer 2: ```suggestion
1. Commit and push your changes to the new branch.
```

> 4. Run the pipeline to deploy the changes. The pipeline will automatically run `terraform plan` and `terraform apply` with the approval to deploy the changes to your environment.

Reviewer 2: ```suggestion
1. Raise a pull request from your branch to the main branch and allow the continuous integration pipeline to run.
1. Once the pipeline has completed, you can open it to examine the plan that was generated and review and changes.
```

> 5. Review the changes in feature branch and ensure everything is working as expected. then merge the changes into the main branch.

Reviewer 2: ```suggestion
1. Review the plan and the changes in feature branch and ensure everything is working as expected, then merge the changes into the main branch.
```

> 6. Run the pipeline again to see the plan and deploy the changes to the Azure environment.

Reviewer 2: ```suggestion
1. The pipeline will trigger from the `main` branch, and run a `plan`. You'll then need to approve the `apply`.
```

### Github
1. In your Github repository, create a new branch to test the upgrade. This ensures your changes are isolated and can be tested without affecting the main branch.

> 2. Modify the `version` field in Terraform module configuration to the desired version. Use version constraints to pin to a specific version or allow automatic upgrades to the latest patch release.

Reviewer 2: ```suggestion
1. Modify the `version` field in Terraform module configuration to the desired version. Use version constraints to pin to a specific version or allow automatic upgrades to the latest patch release.
```

> 

Reviewer 2: ```suggestion
```

> 3. Commit and Sync your changes to the new branch.

Reviewer 2: ```suggestion
1. Commit and push your changes to the new branch.
```

> 4. Run the pipeline to deploy the changes. The pipeline will automatically run `terraform plan` and `terraform apply` with the approval to deploy the changes to your environment.

Reviewer 2: ```suggestion
1. Raise a pull request from your branch to the main branch and allow the continuous integration workflow to run.
1. Once the workflow has completed, you can open it to examine the plan that was generated and review and changes.
```

5. Review the changes in feature branch and ensure everything is working as expected. then merge the changes into the main branch.

> 6. Run the pipeline again to see the plan and deploy the changes to the Azure environment.

Reviewer 2: ```suggestion
1. The workflow will trigger from the `main` branch, and run a `plan`. You'll then need to approve the `apply`.
```
